src/openms/include/OpenMS/CHEMISTRY/VarMod.h

```diff
@@ -0,0 +1,67 @@
+// --------------------------------------------------------------------------
+//                   OpenMS -- Open-Source Mass Spectrometry
+// --------------------------------------------------------------------------
+// Copyright The OpenMS Team -- Eberhard Karls University Tuebingen,
+// ETH Zurich, and Freie Universitaet Berlin 2002-2022.
+//
+// This software is released under a three-clause BSD license:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Amit Kumar Mishra $
+// $Authors: Amit Kumar Mishra $
+// --------------------------------------------------------------------------
+
+#pragma once
+
+#include <OpenMS/CHEMISTRY/ResidueModification.h>
+#include <OpenMS/DATASTRUCTURES/ListUtils.h>
+#include <OpenMS/DATASTRUCTURES/String.h>
+
+#include <set>
+
+namespace OpenMS
+{
+  class OPENMS_DLLAPI VarMod
```

Comment:
In my opinion, the design of this class should be improved. 
Let's discuss a few of your design decisions:
First: why is a class with state needed in the first place? You have some rather complex code and state to merge the residue information.

Wouldn't it be better/possible to structure the code into a static function **groupMergable**
that returns a ```vector<set<ResidueModification*>>``` and takes a ```vector<ResidueModification*>``` as input?

And then a static ```getCometStringFromMergedModification``` that takes one of these ```set<ResidueModification*>``` and returns the string?
That way you have:
-  maximally reduced state (no class members, only static methods). 
- You would also remove dependencies to implementation details (you would have one separate function getCometStringFromMergedModification for exporting) and you can put this search engine specific implementation details of exporting a config string into the namespace or implementation of the Adapter.

Reply:
I don't remember all details but I think the only purpose in the universe wasn't just to get unique modifications but there were some conditions, given satisfied, two same modifications could be merged. That's why there was `isMergeableWith(const VarMod& otherMod)` class and then they were merged.

Reply:
E.g. what if mergability differs between search engines? Then you would need to introduce a strategy pattern into your class etc. … as I see it the two customization points are how merging is performed and how the search config (string) is generated. Implementation with free/static functions and potentially passing a predicate for mergability is in my view a better design.